### ISSUE TITLE
Whitelisting Query Parameters, A "Cache-buster" Trigger, or No Caching When Logged In

### ISSUE THREAD

**Issue by lmaucoin** [putyourlightson/craft-blitz#272]
Hi, guys! Thanks so much for the hard work on the Blitz plugin. It's saved me a lot of headaches! 

Hopefully, the below request makes sense:

### Problem
Visitors to our site often come in with campaign tracking query strings. Very few URLs on our site return unique content based on query strings, but these (often search) parameters are known to our development team. Meanwhile, campaign parameters may go unknown or be unannounced to us, and traffic is sent to our site that doesn't return a cached version of the page, and can result in slow loading.

If we were to treat all URLs with query strings as the same, this would not allow our content editors to see the changes they make as they are making edits. While, ideally, they'd be working on a staging server or using Craft's "Preview" feature, I'm not in control of their editing process, and would like them to have the ability to see non-cached versions of pages on demand.

### Proposed Solution
Similar to the issue in  #177 (requesting to blacklist query parameters), I'd ideally like to see either a whitelist of query parameters that will generate uncached pages. That way, we can set the `2: Cache URLs with query strings as the same page` option, and any URL with a query parameter in the whitelist will generate for the visitor a fresh version of the page.

Another option that only affects our content editors (but doesn't address the visitors) would be a singular "cache buster" trigger option, where passing the query parameter (such as `?nocache=true`) would result in a fresh version of the page being generated for the user.

Something similar to the above solution for our content editors would be to give the option for logged in users to never see cached versions of the page. This sometimes works for me when I'm logged in, but I can't get it to consistently appear when logged in as an editor.

In other words:
- **Whitelist for Query Parameters** to generate an uncached version of a page
- **A "Cache-buster" Trigger** to bypass the cache on a page when used in a query string, or
- **No Cache for Logged In Users Option** for users working on a site to see non-cached versions of site pages.

### Alternatives
While I could use the `excludedQueryStringParams` option with `'queryStringCaching' => 1`, we would see the following issues:
- Maintaining a list of campaign parameters that the development team may be unaware of.
- The generation (and maintenance?) of cache for every non-blacklisted key/value pair of query strings could grow out of control.

Additionally, if only for our content editors' sake, I could setup an easy way for them to refresh the cache on a given page they are looking at via the API, but I fear it would be more complicated to explain to them than something like a cache-buster, and will result in multiple background jobs to generate cache for that page.

### Additional context
For search items, the URI for anything that utilizes the query parameters can certainly be added to the Excluded URIs list. When used in conjunction with the `Cache URLs with query strings as the same page` setting, this works for most of our cases, but leaves our content editors unable to consistently view their edits without forcing a refresh of the cache.

**Comment by bencroker** [comment]
Glad to hear you're enjoying Blitz!

> If we were to treat all URLs with query strings as the same, this would not allow our content editors to see the changes they make as they are making edits. While, ideally, they'd be working on a staging server or using Craft's "Preview" feature, I'm not in control of their editing process, and would like them to have the ability to see non-cached versions of pages on demand.

The live preview feature should continue to work regardless, even if a page is already cached. Are you seeing something different, and if so then what version of the plugin and Craft are you running?

**Comment by lmaucoin** [comment]
Thanks, @bencroker! The Live Preview seems like a good solution for our team of editors if we can train them on its use, and this may also be a case of me pulling the classic developer move in overengineering a solution. 😅

I will close this issue out, but if having a whitelist of query parameters for generating unique content is ever down the road, I would definitely enjoy it.

**Comment by mrw** [comment]
Hello @bencroker! I'd also like to request a whitelist of query string parameters, for the same reason as mentioned by @lmaucoin: 

> Very few URLs on our site return unique content based on query strings, but these (often search) parameters are known to our development team. Meanwhile, campaign parameters may go unknown or be unannounced to us, and traffic is sent to our site that doesn't return a cached version of the page, and can result in slow loading.

We know the 2-3 query parameters that should impact page rendering and would like to specifically set them. But then we'd like blitz to ignore everything else, because it could be random from a spider, a campaign parameter, etc.. Is that possible?

**Comment by bencroker** [comment]
I understand your point of view Matt, but we discussed this internally in the past and came up with the following counter arguments that led us to our original conclusion.

If a spider visits a page using a random query string `domain.com/?name=spidy` and URLs with query strings are set to _not_ be cached, then each time spidy hits that URL, the web server will have to render the page with PHP. If on the other hand, URLs with query strings are set to be cached, the web server will have to render the page with PHP only once, as all subsequent requests will be served the cached version. So in the latter case, spidy can now hit that URL as many times as it wants without negatively impacting the server.

Of course spidy could start hitting multiple random query strings. If URLs with query strings are set to be cached _as unique pages_, then this would result in multiple spidy pages being cached, but this is less of a concern than the server being bogged down by the first scenario above.

For the case of campaign parameters, they can be added in the `excludedQueryStringParams` config setting, assuming they are known in advance (they should be if the campaigns are genuine).

**Comment by mrw** [comment]
Hi Ben! I totally get it. If it's helpful though, I just want to address this from your note to help explain our use case:

> Of course spidy could start hitting multiple random query strings. If URLs with query strings are set to be cached as unique pages, then this would result in multiple spidy pages being cached, but this is less of a concern than the server being bogged down by the first scenario above.

One of our big frequent goals of caching with Blitz is to significantly reduce server load on resource-heavy pages, and lessen the ability for a site to get DoSed by a spider. Your scenario is less of a concern in the sense that Redis/file caching is abundant and we could probably store tens of thousands of unique URLs. However, it's still a big issue in the other sense. Without the ability to whitelist query strings, spidy can hit us with thousands of requests to random query strings and cause very resource-heavy pages to be rendered thousands of times, effectively DoSing a site. With a query string whitelist we can make sure that doesn't happen. Calls to `?sort=asc&spidy1=yes` use the exact same cache as `?sort=asc&spidy2=yes`.

I hope that helps explain this feature request a bit more. Thank you for your consideration!

**Comment by bencroker** [comment]
The reason I am resistant to adding a whitelist is that it will not work with the existing cache URL strategy options (nor with server rewrites) and I would rather not change this in a minor plugin version. I am considering adding the ability to use a regular expression in the `excludedQueryStringParams` config setting, so for example both of the following would be possible:

```
// Array of query string param names
'excludedQueryStringParams' => ['gclid', 'fbclid'],

// Regular expression string: exclude any query string params that are not `sort` or `search`
'excludedQueryStringParams' => '^(?!sort$|search$).*',
```

It's not particularly elegant but it works. Any thoughts on this approach?

**Comment by mrw** [comment]
That sounds like a great approach that would solve this problem. Thank you!

**Comment by bencroker** [comment]
Added in https://github.com/putyourlightson/craft-blitz/commit/5a987ef45b040dbffb1d452b086973283eb9bebe for the next release.

The value of `excludedQueryStringParams` must be an array, so this is how you would exclude any query string param with a name that is neither `sort` nor `search`:
```
'excludedQueryStringParams' => ['^(?!sort$|search$).*']
```

Note that if using server redirects, you'll need to set up your own rules to ignore these excluded query string params.

**Comment by mrw** [comment]
This is great, thank you so much!

**Comment by bencroker** [comment]
Released in 3.8.0.

**Comment by bencroker** [comment]
FYI Blitz 3.10.0 added  a new `includedQueryStringParams` config setting that should make specifying query string params to include easier: https://github.com/putyourlightson/craft-blitz/blob/c48ad3a1806eafcd8c5539c8fc79cc3ab6e9db85/src/config.php#L148-L149
